File: core/zig/src/random.zig
pub const XorshiftResult = struct {
    state: u32,
    value: f64,
};

pub const RandomSliceResult = struct {
    state: u32,
    slice: i32,
};

pub fn xorshift32(state: u32) XorshiftResult {
    var next = state;
    next ^= next << 13;
    next ^= next >> 17;
    next ^= next << 5;
    return .{
        .state = next,
        .value = @as(f64, @floatFromInt(next)) / @as(f64, @floatFromInt(0x1_0000_0000)),
    };
}

pub fn randomSlice(state: u32, slice_count: u32, current_slice: i32) RandomSliceResult {
    const random = xorshift32(state);
    var slice: i32 = @intFromFloat(@floor(random.value * @as(f64, @floatFromInt(slice_count))));
    if (slice_count > 1 and slice == current_slice) {
        slice = @rem(slice + 1, @as(i32, @intCast(slice_count)));
    }
    return .{
        .state = random.state,
        .slice = slice,
    };
}

File: core/rust/src/random.rs
pub fn xorshift32(state: u32) -> (u32, f64) {
    let mut next = state;
    next ^= next.wrapping_shl(13);
    next ^= next.wrapping_shr(17);
    next ^= next.wrapping_shl(5);
    let value = f64::from(next) / 4294967296.0;
    (next, value)
}

pub fn random_slice(state: u32, slice_count: u32, current_slice: i32) -> (u32, i32) {
    let (next_state, value) = xorshift32(state);
    let mut slice = (value * f64::from(slice_count)).floor() as i32;
    if slice_count > 1 && slice == current_slice {
        slice = (slice + 1) % slice_count as i32;
    }
    (next_state, slice)
}
